server/sonar-server/src/main/java/org/sonar/server/computation/issue/IssueAssigner.java

```diff
@@ -115,38 +118,43 @@ private String guessScmAuthor(@Nullable Integer line) {
     return reportReader.readChangesets(component.getReportAttributes().getRef());
   }
 
-  private BatchReport.Changesets loadScmChangesetsFromIndex(Component component) {
-    List<SourceLineDoc> lines = sourceLineIndex.getLines(component.getUuid());
-    Map<String, BatchReport.Changesets.Changeset> changesetByRevision = new HashMap<>();
-    BatchReport.Changesets.Builder scmBuilder = BatchReport.Changesets.newBuilder()
-      .setComponentRef(component.getReportAttributes().getRef());
-    for (SourceLineDoc sourceLine : lines) {
-      String scmRevision = sourceLine.scmRevision();
-      if (scmRevision == null || changesetByRevision.get(scmRevision) == null) {
-        BatchReport.Changesets.Changeset.Builder changeSetBuilder = BatchReport.Changesets.Changeset.newBuilder();
-        String scmAuthor = sourceLine.scmAuthor();
-        if (scmAuthor != null) {
-          changeSetBuilder.setAuthor(scmAuthor);
-        }
-        Date scmDate = sourceLine.scmDate();
-        if (scmDate != null) {
-          changeSetBuilder.setDate(scmDate.getTime());
-        }
-        if (scmRevision != null) {
-          changeSetBuilder.setRevision(scmRevision);
-        }
+  private BatchReport.Changesets loadScmChangesetsFromDb(Component component) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      Optional<Iterable<DbFileSources.Line>> lines = sourceService.getLines(dbSession, component.getUuid(), 1, Integer.MAX_VALUE);
+      Map<String, BatchReport.Changesets.Changeset> changesetByRevision = new HashMap<>();
+      BatchReport.Changesets.Builder scmBuilder = BatchReport.Changesets.newBuilder()
+        .setComponentRef(component.getReportAttributes().getRef());
+      if (lines.isPresent()) {
+        for (DbFileSources.Line sourceLine : lines.get()) {
+          String scmRevision = sourceLine.getScmRevision();
+          if (scmRevision == null || changesetByRevision.get(scmRevision) == null) {
+            BatchReport.Changesets.Changeset.Builder changeSetBuilder = BatchReport.Changesets.Changeset.newBuilder();
+            if (sourceLine.hasScmAuthor()) {
+              changeSetBuilder.setAuthor(sourceLine.getScmAuthor());
+            }
+            if (sourceLine.hasScmDate()) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)